### PR TITLE
DataSpace can be included and localized in relational stores

### PR DESCRIPTION
**Pull request by haroonsherjan**
#### What type of PR is this?
Improvement

#### What does this PR do / why is it needed ?

Consumers can include and localize the relational stores under a dataspace using the existing include syntax for mapping and store.

### CODE REVIEW

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/PureModel.java [hunk 1]
```diff
@@ -469,44 +456,11 @@ private void loadDataElements(PureModelContextDataIndex pure)
...
-    }
-
-    public void loadConnectionsAndRuntimes(PureModelContextDataIndex pure)
-    {
-        // Connections must be loaded before runtimes
-        pure.connections.forEach(this::processSecondPass);
-        pure.runtimes.forEach(this::processSecondPass);
-    }
-
     private void loadOtherElementsPreStores(PureModelContextDataIndex pure)
```
Comment (akphi) [inline legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/PureModel.java:459]: I think this kind of changes should go in its own PR so we can easily rollback in case. You would also want to remote run this with the internal codebase where we have more plugins

File: legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/extension/PureGrammarParserExtension.java [hunk 2]
```diff
@@ -61,4 +67,26 @@ default Iterable<? extends MappingIncludeParser> getExtraMappingIncludeParsers()
     {
         return Collections.emptyList();
     }
+
+    static String parseIncludedStoreType(String type, List<Function<String, String>> processors)
```
Comment (akphi) [inline legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/extension/PureGrammarParserExtension.java:71]: Maybe move this out to the MappingParserHelper or something?

File: legend-engine-protocol-pure/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/packageableElement/store/IncludedStore.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.finos.legend.engine.protocol.pure.v1.model.packageableElement.store;
+
+import org.finos.legend.engine.protocol.pure.v1.model.SourceInformation;
+
+public class IncludedStore
```
Comment (akphi) [inline legend-engine-protocol-pure/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/packageableElement/store/IncludedStore.java:19]: `abstract`?

Comment (akphi) [reply legend-engine-protocol-pure/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/packageableElement/store/IncludedStore.java:19]: we probably should do the same as what we did for `includedMapping`

File: legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java [hunk 4]
```diff
@@ -199,4 +180,30 @@ public Map<String, IncludedMappingHandler> getExtraIncludedMappingHandlers()
...
+                "dataspace", (IncludedStore includedStore, CompileContext context) ->
+                {
+                    Root_meta_pure_metamodel_dataSpace_DataSpace dataspace =
+                            DataSpaceCompilerExtension.dataSpacesIndex.get(includedStore.name);
+                    return getDatabasesUnderDataSpace(dataspace, context);
+                }
+        );
+    }
+
+    public static RichIterable<Store> getDatabasesUnderDataSpace(Root_meta_pure_metamodel_dataSpace_DataSpace dataspace, CompileContext context)
```
Comment (akphi) [inline legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java:197]:
```suggestion
    public static RichIterable<Store> getStoresUnderDataSpace(Root_meta_pure_metamodel_dataSpace_DataSpace dataspace, CompileContext context)
```

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/PureModel.java (hunk 1 above)
Comment (akphi) [reply legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/PureModel.java:459]: Also, I'm not even sure if this is the right thing to do generally. This means somebody who's going to add a new DSL need to know about the implementation of the processing passes of a lot of other plugins. Is that really what we want?

**Review by akphi — COMMENTED** [review]
Just a partial review, I left some comments

File: legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java (hunk 4 above)
Comment (haroonsherjan) [reply legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java]: Done

File: legend-engine-protocol-pure/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/packageableElement/store/IncludedStore.java (hunk 3 above)
Comment (haroonsherjan) [reply legend-engine-protocol-pure/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/packageableElement/store/IncludedStore.java]: Makes sense. Done.

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/extension/CompilerExtensions.java [hunk 5]
```diff
@@ -468,4 +473,16 @@ public IncludedMappingHandler getExtraIncludedMappingHandlers(String classType)
     {
         return this.extraIncludedMappingHandlers.get(classType);
     }
+
+    public Function2<IncludedStore, CompileContext, RichIterable<Store>> getExtraIncludedStoreHandlers(IncludedStore includedStore)
+    {
+        if (this.extraIncludedStoreHandlers.get(includedStore.type) == null)
```
[the remark this answers is not in the document]
Comment (haroonsherjan) [reply legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/extension/CompilerExtensions.java]: Makes sense. Done

File: legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/extension/PureGrammarParserExtension.java (hunk 2 above)
Comment (haroonsherjan) [reply legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/extension/PureGrammarParserExtension.java]: Makes sense. Done

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/PureModel.java (hunk 1 above)
Comment (rafaelbey) [reply legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/PureModel.java:459]: Agree this is something very sensitive, and we might want to change on a separate PR before we make further changes that depend on this change.

As for the knowledge people need to have - I think that is true regardless of this implementation.  What we don't have is clear documentation into what should go on each phase and what the core part of the platform does on each phase.  Like, if I create an element that depends on Store, I need to know what I should do on each compilation phase  of my component, specially if it needs Store to be on specific state.

Comment (haroonsherjan) [reply legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/PureModel.java]: Solved this without re-ordering compilation

**Comment by finos-admin** [comment]
This PR was closed because it has been inactive for 35 days. Please re-open if this PR is still relevant.
